src/unix/core.c

```diff
@@ -1517,3 +1526,68 @@ void uv_sleep(unsigned int msec) {
 
   assert(rc == 0);
 }
+
+
+void uv__metrics_update_idle_time(uv_loop_t* loop) {
+  uv__metrics_loop_t* mloop;
+  uint64_t entry_time;
+  uint64_t idle_time;
+  uint64_t exit_time;
+
+  if (!(loop->flags & UV_LOOP_IDLE_TIME))
+    return;
+
+  mloop = uv__metrics_get_mloop(loop);
+
+  /* The thread running uv__metrics_update_idle_time() is always the same
+   * thread that sets provider_entry_time. So it's unnecessary to perform the
+   * atomic operation retrieving this value.
+   * Reason for this check is in case this function has been called a second
+   * time. Since it's always called just after uv__io_poll() to make sure the
+   * idle time has been correctly calculated.
+   */
+  if (mloop->provider_entry_time == 0)
+    return;
+
+  /* Grab the exit_time before starting the atomic operations. */
+  exit_time = uv_hrtime();
+
+  /* It's important to run the operations in this order. It will prevent
+   * accidentally reporting incorrect values and removes the need for a lock.
+   */
+  UV_ATOMIC_EXCHANGE(&mloop->provider_entry_time, &entry_time, 0);
+  UV_ATOMIC_ADD_FETCH(&mloop->idle_time, &idle_time, exit_time - entry_time);
+}
+
+
+void uv__metrics_set_provider_entry_time(uv_loop_t* loop) {
+  uv__metrics_loop_t* mloop;
+  uint64_t now;
+
+  if (!(loop->flags & UV_LOOP_IDLE_TIME))
+    return;
+
+  now = uv_hrtime();
+  mloop = uv__metrics_get_mloop(loop);
+  UV_ATOMIC_STORE(&mloop->provider_entry_time, now);
+}
+
+
+uv__metrics_loop_t* uv__metrics_get_mloop(uv_loop_t* loop) {
+  return (uv__metrics_loop_t*)loop->active_reqs.unused[1];
+}
+
+
+uint64_t uv_metrics_idle_time(uv_loop_t* loop) {
+  uv__metrics_loop_t* mloop;
+  uint64_t entry_time;
+  uint64_t idle_time;
+
+  mloop = uv__metrics_get_mloop(loop);
+  UV_ATOMIC_LOAD(&mloop->idle_time, &idle_time);
+  UV_ATOMIC_LOAD(&mloop->provider_entry_time, &entry_time);
+
+  if (entry_time > 0)
+    idle_time += uv_hrtime() - entry_time;
+  return idle_time;
+}
```

Comment:
This should handle the case where `UV_LOOP_IDLE_TIME` is not set by returning `UV_EINVAL`. Maybe changing the signature to something like: `int uv_metrics_idle_time(uv_loop_t* loop, uint64_t* idle_time)` might work.

Reply:
I think simply returning `0` is enough. The API documentation says it only starts counting once the flag has been set, so IMO it's within reason that no time has been collected before being set.